engine/crates/parser-sdl/src/validations.rs

```diff
@@ -315,3 +319,25 @@ fn validate_federation_joins(registry: &Registry) -> Vec<RuleError> {
 
     errors
 }
+
+pub fn validate_federation_keys(pos: Pos, key: &KeyDirective, object: &ObjectType) -> Vec<RuleError> {
+    let mut errors = Vec::new();
+
+    errors.extend(key.validate().into_iter().map(|error| RuleError::new(vec![pos], error)));
+
+    for field in &key.fields.0 .0 {
+        if object.field_by_name(&field.field).is_none() {
+            errors.push(RuleError::new(
+                vec![pos],
+                format!(
+                    "The object {} has a key that requires the field {} but that field isn't present",
+                    object.name, &field.field,
+                ),
+            ));
+        }
+        // In an ideal world we'd also validate any nested keys, but we don't really have
+        // access to any of the other types involved at this point :(
```

Comment:
Nested keys seem super rare anyway

Reply:
Oh, this comment isn't even true now that I've moved the validation...

Although I don't really want to write nested key validation just now, so maybe I'll just fix the comment 😅